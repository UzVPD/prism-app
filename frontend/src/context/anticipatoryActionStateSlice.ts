import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { LayerDefinitions } from 'config/utils';
import { AnticipatoryActionLayerProps } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from './store';

export const AAlayerKey = 'anticipatory_action';

const AACSVKyes: [string, string][] = [
  ['Category', 'category'],
  ['District', 'district'],
  ['Index', 'index'],
  ['Month', 'month'],
  ['Phase', 'phase'],
  ['Probability', 'probability'],
  ['Trigger', 'trigger'],
  ['Trigger_nb', 'triggerNB'],
  ['Trigger_type', 'triggerType'],
  ['Type', 'type'],
  ['Windows', 'windows'],
  ['Year_of_issue', 'yearOfIssue'],
];

function transform(data: any[], keys: [string, string][]) {
  return data
    .filter(x => !!x[keys[0][0]]) // filter empty rows
    .map(obj => {
      const entries = keys.map(k => [k[1], obj[k[0]]]);
      const month = obj.Month.padStart(2, '0');
      const year = obj.Year_of_issue.split('-')[0];
      const date = `${year}-${month}-01`;
      return Object.fromEntries([...entries, ['date', date]]);
    });
}

export interface AnticipatoryActionData {
  category: 'Leve' | 'Moderado' | 'Severo';
  district: string;
  index: string;
  month: string;
  phase: 'Ready' | 'Set';
  probability: string;
  trigger: string;
  triggerNB: string;
  triggerType: string;
  type: string;
  windows: string;
  yearOfIssue: string;
  date: string;
}

type AnticipatoryActionState = {
  data: AnticipatoryActionData[];
  loading: boolean;
  error: string | null;
};

const initialState: AnticipatoryActionState = {
  data: [],
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  { rows: AnticipatoryActionData[]; columns: string[] },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const layer = LayerDefinitions[AAlayerKey] as AnticipatoryActionLayerProps;

  const url = layer.baseUrl;
  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results =>
        resolve({
          rows: transform(results.data, AACSVKyes),
          columns: Object.keys(results.data[0]),
        }),
      error: error => reject(error),
    }),
  );
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {},
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.rows,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AnticipatoryActionDataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

// export actions
// export const {  } = tableStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
