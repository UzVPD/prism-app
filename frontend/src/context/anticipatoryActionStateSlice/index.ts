import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { DateItem, DatesPropagation, Validity } from 'config/types';
import { generateIntermediateDateItemFromValidity } from 'utils/server-utils';
import { appConfig } from 'config';
import { AAWindowKeyToLayerId, AAWindowKeys } from 'config/utils';
import { AADataSeverityOrder } from 'components/MapView/LeftPanel/AnticipatoryActionPanel/utils';
import type { CreateAsyncThunkTypes, RootState } from '../store';
import {
  AACategoryType,
  AAPhaseType,
  AnticipatoryActionData,
  AnticipatoryActionDataRow,
  AnticipatoryActionState,
  allWindowsKey,
} from './types';

const AACSVKeys: string[] = [
  'district',
  'index',
  'category',
  'window',
  'season',
  'type',
  'date_ready',
  'date_set',
  'issue_ready',
  'issue_set',
  'prob_ready',
  'prob_set',
  'trigger_ready',
  'trigger_set',
  'state',
  'new_tag',
];

const sortFn = (a: AnticipatoryActionDataRow, b: AnticipatoryActionDataRow) => {
  const aDate = new Date(a.date).valueOf();
  const bDate = new Date(b.date).valueOf();
  if (aDate > bDate) {
    return -1;
  }
  if (bDate > aDate) {
    return 1;
  }
  const aSev = AADataSeverityOrder(a.category, a.phase);
  const bSev = AADataSeverityOrder(b.category, b.phase);
  if (aSev > bSev) {
    return -1;
  }
  if (bSev > aSev) {
    return 1;
  }
  return 0;
};

function transform(data: any[]) {
  const nonEmpty = data.filter(x => !!x[AACSVKeys[0]]); // filter empty rows
  const parsed = nonEmpty
    .map(x => {
      const common = {
        category: x.category,
        district: x.district,
        index: x.index,
        type: x.type,
        window: x.window,
        new: Boolean(Number(x.new_tag)),
      };

      const ready = {
        phase: 'Ready',
        probability: Number(x.prob_ready),
        trigger: Number(x.trigger_ready),
        date: x.date_ready,
      };

      const set = {
        phase: 'Set',
        probability: Number(x.prob_set),
        trigger: Number(x.trigger_set),
        date: x.date_set,
      };

      return [
        { ...common, ...ready },
        { ...common, ...set },
      ];
    })
    .flat() as AnticipatoryActionDataRow[];

  const groupedByWinDistIndexMap = new Map<
    string,
    AnticipatoryActionDataRow[]
  >();
  parsed.forEach(x => {
    const key = `${x.window}_${x.district}_${x.index}`;
    const val = groupedByWinDistIndexMap.get(key);
    groupedByWinDistIndexMap.set(key, val ? [...val, x] : [x]);
  });
  const extraRows = Array.from(groupedByWinDistIndexMap.values())
    .map(x => {
      // eslint-disable-next-line fp/no-mutating-methods
      const sorted = x.sort((a, b) => -sortFn(a, b));
      let isSetSev: boolean = false;
      let isSetMod: boolean = false;
      let isSetMil: boolean = false;
      return sorted.reduce((acc, curr) => {
        const prev = acc.length > 0 ? acc[acc.length - 1] : undefined;
        if (curr.probability > curr.trigger && curr.phase === 'Set') {
          if (curr.category === 'Severe') {
            // eslint-disable-next-line fp/no-mutation
            isSetSev = true;
          } else if (curr.category === 'Moderate') {
            // eslint-disable-next-line fp/no-mutation
            isSetMod = true;
          } else {
            // eslint-disable-next-line fp/no-mutation
            isSetMil = true;
          }
        }
        if (
          (isSetSev || isSetMod || isSetMil) &&
          (!prev || prev.date !== curr.date)
        ) {
          let newElems: AnticipatoryActionDataRow[] = [];
          if (isSetMil) {
            // eslint-disable-next-line fp/no-mutation
            newElems = [
              ...newElems,
              { ...curr, computedRow: true, category: 'Mild', phase: 'Set' },
            ];
          }
          if (isSetMod) {
            // eslint-disable-next-line fp/no-mutation
            newElems = [
              ...newElems,
              {
                ...curr,
                computedRow: true,
                category: 'Moderate',
                phase: 'Set',
              },
            ];
          }
          if (isSetSev) {
            // eslint-disable-next-line fp/no-mutation
            newElems = [
              ...newElems,
              { ...curr, computedRow: true, category: 'Severe', phase: 'Set' },
            ];
          }
          return [...acc, ...newElems];
        }
        return acc;
      }, [] as AnticipatoryActionDataRow[]);
    })
    .flat();

  const extended = [...parsed, ...extraRows];

  const validity: Validity = {
    mode: DatesPropagation.DEKAD,
    forward: 3,
  };

  const monitoredDistricts = [...new Set(extended.map(x => x.district))];
  const emptyDistricts = Object.fromEntries(
    monitoredDistricts.map(x => [x, [] as AnticipatoryActionDataRow[]]),
  );

  const windowData = AAWindowKeys.map(windowKey => {
    const filtered = extended.filter(x => x.window === windowKey);

    // eslint-disable-next-line fp/no-mutating-methods
    const dates = [
      ...new Set(filtered.map(x => new Date(x.date).getTime())),
    ].sort();
    const availableDates = generateIntermediateDateItemFromValidity(
      dates,
      validity,
    );

    const groupedByDistrict = new Map<string, AnticipatoryActionDataRow[]>();
    filtered.forEach(x => {
      const { district } = x;
      const rows = groupedByDistrict.get(district);
      groupedByDistrict.set(district, rows ? [...rows, x] : [x]);
    });

    const result = Object.fromEntries(
      Array.from(groupedByDistrict.entries()).map(x => [
        x[0],
        // eslint-disable-next-line fp/no-mutating-methods
        x[1].sort(sortFn),
      ]),
    );

    return {
      data: { ...emptyDistricts, ...result },
      availableDates,
      windowKey,
    };
  });

  return { windowData, monitoredDistricts };
}

const emptyWindows = {
  'Window 1': {},
  'Window 2': {},
} as AnticipatoryActionState['renderedDistricts'];

const initialState: AnticipatoryActionState = {
  data: {},
  availableDates: undefined,
  filters: {
    selectedDate: undefined,
    selectedWindow: allWindowsKey,
    categories: {
      Severe: true,
      Moderate: true,
      Mild: true,
      na: true,
      ny: true,
    },
  },
  renderedDistricts: emptyWindows,
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    windowData: {
      data: AnticipatoryActionData;
      availableDates: DateItem[];
      windowKey: typeof AAWindowKeys[number];
    }[];
    monitoredDistricts: string[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const url = appConfig.anticipatoryActionUrl;

  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(transform(results.data)),
      error: error => reject(error),
    }),
  );
});

interface CalculateMapRenderedDistrictsParams {
  filters: AnticipatoryActionState['filters'];
  data: AnticipatoryActionState['data'];
}

function calculateMapRenderedDistricts({
  filters,
  data,
}: CalculateMapRenderedDistrictsParams) {
  const { selectedDate, categories } = filters;
  const res = Object.entries(data)
    .map(([winKey, districts]) => {
      if (!districts) {
        return undefined;
      }
      const entries = Object.entries(districts).map(
        ([districtName, districtData]) => {
          const dateData = districtData.filter(x => x.date === selectedDate);
          if (dateData.length === 0) {
            return [
              districtName,
              { category: 'ny', phase: 'ny', isNew: false },
            ];
          }

          const validData = dateData.filter(
            x =>
              (x.computedRow ||
                (!Number.isNaN(x.probability) && x.probability >= x.trigger)) &&
              categories[x.category],
          );
          if (validData.length === 0) {
            return [
              districtName,
              { category: 'na', phase: 'na', isNew: false },
            ];
          }

          const max = validData.reduce((m, curr) => {
            if (
              AADataSeverityOrder(curr.category, curr.phase) >
              AADataSeverityOrder(m.category, m.phase)
            ) {
              return curr;
            }
            return m;
          }, validData[0]);
          return [
            districtName,
            {
              category: max.category,
              phase: max.phase,
              isNew: max.computedRow ? false : max.new,
            },
          ];
        },
      );
      return [
        winKey,
        Object.fromEntries(entries) as {
          [district: string]: {
            category: AACategoryType;
            phase: AAPhaseType;
            isNew: boolean;
          };
        },
      ];
    })
    .filter((x): x is any[] => x !== undefined);

  const obj = Object.fromEntries(res) as {
    [windowKey: string]: {
      [district: string]: {
        category: AACategoryType;
        phase: AAPhaseType;
        isNew: boolean;
      };
    };
  };

  return { ...emptyWindows, ...obj };
}

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setAAFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<{
          selectedDate: string | undefined;
          selectedWindow: typeof AAWindowKeys[number] | typeof allWindowsKey;
          categories: Partial<Record<AACategoryType, boolean>>;
        }>
      >,
    ) => {
      const { categories, ...rest } = payload;
      const newCategories =
        categories !== undefined
          ? { ...state.filters.categories, ...categories }
          : state.filters.categories;
      const newFilters = {
        ...state.filters,
        ...rest,
        categories: newCategories,
      };
      return {
        ...state,
        filters: newFilters,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: newFilters,
          data: state.data,
        }),
      };
    },
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.data]),
      ),
      availableDates: Object.fromEntries(
        payload.windowData.map(x => [
          AAWindowKeyToLayerId[x.windowKey],
          x.availableDates,
        ]),
      ),
      monitoredDistricts: payload.monitoredDistricts,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AADataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AAAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AAFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.filters;

export const AARenderedDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.renderedDistricts;

// export actions
export const { setAAFilters } = anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
