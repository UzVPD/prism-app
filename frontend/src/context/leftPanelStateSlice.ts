import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { appConfig } from 'config';
import type { RootState } from './store';

const { hidePanel } = appConfig;

export enum Panel {
  None = 'none',
  Layers = 'layers',
  Charts = 'charts',
  Analysis = 'analysis',
  Tables = 'tables',
}

type LeftPanelState = {
  tabValue: Panel;
};

const initialState: LeftPanelState = {
  tabValue: hidePanel ? Panel.None : Panel.Layers,
};

export const leftPanelSlice = createSlice({
  name: 'leftPanelState',
  initialState,
  reducers: {
    setTabValue: (state, { payload }: PayloadAction<Panel>) => {
      if (payload === state.tabValue) {
        return {
          ...state,
          tabValue: Panel.None,
        };
      }
      return {
        ...state,
        tabValue: payload,
      };
    },
  },
});

// Getters
export const leftPanelTabValueSelector = (state: RootState): Panel =>
  state.leftPanelState.tabValue;

// Setters
export const { setTabValue } = leftPanelSlice.actions;

export default leftPanelSlice.reducer;
