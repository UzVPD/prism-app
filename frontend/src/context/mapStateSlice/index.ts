import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Map as MapBoxMap } from 'mapbox-gl';
import { LayerKey, LayerType, LayerForm } from '../../config/types';
import { LayerDefinitions } from '../../config/utils';
import { LayerData, LayerDataTypes, loadLayerData } from '../layers/layer-data';

interface DateRange {
  startDate?: number;
  // TODO this field is updated, but doesn't seem to be used yet
  endDate?: number;
}

export type MapState = {
  layers: LayerType[];
  dateRange: DateRange;
  mapboxMap: MapGetter;
  errors: string[];
  // TODO this shouldn't be any
  layersData: LayerData<any>[];
  layerForms: LayerForm[];
  // Keep track of layer id which are currently loading its layerData.
  // Note that layerData is mainly for storing vector map data.
  // Tile image loading for raster layer is tracked separately on mapTileLoadingStateSlice
  loadingLayerIds: LayerKey[];
};

// MapboxGL's map type contains some kind of cyclic dependency that causes an infinite loop in immers's change
// tracking. To save it off, we wrap it in a JS closure so that Redux just checks the function for changes, rather
// than recursively walking the whole object.
type MapGetter = () => MapBoxMap | undefined;

const initialState: MapState = {
  layers: [],
  dateRange: {} as DateRange,
  mapboxMap: (() => {}) as MapGetter,
  errors: [],
  layersData: [],
  layerForms: [],
  loadingLayerIds: [],
};

export type FormInputChange = {
  layerId: string;
  inputId: string;
  value: string;
};

function keepLayer(layer: LayerType, payload: LayerType) {
  // Simple function to control which layers can overlap.
  return (
    payload.id !== layer.id &&
    (payload.type !== layer.type || payload.type === 'boundary')
  );
}

export const mapStateSlice = createSlice({
  name: 'mapState',
  initialState,
  reducers: {
    addLayer: (
      { layers, layerForms, ...rest },
      { payload }: PayloadAction<LayerType>,
    ) => {
      const layersToAdd = payload?.group?.activateAll
        ? Object.values(LayerDefinitions).filter(l =>
            payload?.group?.layers?.map(layer => layer.id).includes(l.id),
          )
        : [payload];

      const filteredLayers = layers.filter(layer => keepLayer(layer, payload));

      // Keep boundary layers at the top of our stack
      const newLayers =
        payload.type === 'boundary'
          ? [...layersToAdd, ...filteredLayers]
          : [...filteredLayers, ...layersToAdd];

      const layerForm =
        'formInputs' in payload
          ? { id: payload.id, inputs: payload.formInputs! }
          : null;

      return {
        ...rest,
        layers: newLayers,
        layerForms: layerForms.concat(layerForm ? [layerForm] : []),
      };
    },
    removeLayerData: (
      { layersData, layerForms, ...rest },
      { payload }: PayloadAction<LayerType>,
    ) => ({
      ...rest,
      layersData: layersData.filter(({ layer }) => layer.id !== payload.id),
      layerForms: layerForms.filter(({ id }) => id !== payload.id),
    }),

    removeLayer: (
      { layers, ...rest },
      { payload }: PayloadAction<LayerType>,
    ) => ({
      ...rest,
      layers: layers.filter(({ id }) =>
        // Keep layers without group and layers with group from different group.
        payload.group
          ? !payload.group.layers.map(l => l.id).includes(id)
          : id !== payload.id,
      ),
    }),

    setFormInputValue: (
      { layerForms, ...rest },
      { payload }: PayloadAction<FormInputChange>,
    ) => ({
      ...rest,
      layerForms: layerForms.map(layerForm =>
        layerForm.id === payload.layerId
          ? {
              id: layerForm.id,
              inputs: layerForm.inputs.map(({ id, value, ...restInput }) => ({
                id,
                value: id === payload.inputId ? payload.value : value,
                ...restInput,
              })),
            }
          : layerForm,
      ),
    }),

    updateDateRange: (state, { payload }: PayloadAction<DateRange>) => ({
      ...state,
      dateRange: payload,
    }),

    setMap: (state, { payload }: PayloadAction<MapGetter>) => ({
      ...state,
      mapboxMap: payload,
    }),

    dismissError: (
      { errors, ...rest },
      { payload }: PayloadAction<string>,
    ) => ({
      ...rest,
      errors: errors.filter(msg => msg !== payload),
    }),

    updateLayerOpacity: (
      { layers, ...rest },
      { payload }: PayloadAction<Pick<LayerType, 'id' | 'opacity'>>,
    ) => ({
      ...rest,
      layers: layers.map(layer => ({
        ...layer,
        opacity: layer.id === payload.id ? payload.opacity : layer.opacity,
      })),
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      loadLayerData.fulfilled,
      (
        { layersData, loadingLayerIds, ...rest },
        { payload }: PayloadAction<LayerDataTypes>,
      ) => ({
        ...rest,
        loadingLayerIds: loadingLayerIds.filter(id => id !== payload.layer.id),
        layersData: layersData.concat(payload),
      }),
    );

    builder.addCase(
      loadLayerData.rejected,
      ({ loadingLayerIds, errors, ...rest }, action) => ({
        ...rest,
        loadingLayerIds: loadingLayerIds.filter(
          id => id !== action.meta.arg.layer.id,
        ),
        errors: errors.concat(
          action.error.message ? action.error.message : action.error.toString(),
        ),
      }),
    );

    builder.addCase(
      loadLayerData.pending,
      ({ loadingLayerIds, ...rest }, action) => ({
        ...rest,
        loadingLayerIds: loadingLayerIds.concat([action.meta.arg.layer.id]),
      }),
    );
  },
});

// Setters
export const {
  addLayer,
  removeLayer,
  setFormInputValue,
  updateDateRange,
  setMap,
  // TODO unused
  updateLayerOpacity,
  removeLayerData,
} = mapStateSlice.actions;

export default mapStateSlice.reducer;
