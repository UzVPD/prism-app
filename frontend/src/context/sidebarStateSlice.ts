import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from './store';

type SidebarState = {
  tabValue: number;
};

const initialState: SidebarState = {
  tabValue: 0,
};

export const sidebarSlice = createSlice({
  name: 'sidebarState',
  initialState,
  reducers: {
    setTabValue: (state, { payload }: PayloadAction<number>) => ({
      ...state,
      tabValue: payload,
    }),
  },
});

// Getters
export const sidebarTabValueSelector = (state: RootState): number =>
  state.sidebarState.tabValue;

// Setters
export const { setTabValue } = sidebarSlice.actions;

export default sidebarSlice.reducer;
