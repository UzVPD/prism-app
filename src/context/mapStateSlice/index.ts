import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Map as MapBoxMap } from 'mapbox-gl';
import { LayerType, LayerForm } from '../../config/types';
import { LayerDefinitions } from '../../config/utils';
import { LayerData, LayerDataTypes, loadLayerData } from '../layers/layer-data';

interface DateRange {
  startDate?: number;
  // TODO this field is updated, but doesn't seem to be used yet
  endDate?: number;
}

export type MapState = {
  layers: LayerType[];
  dateRange: DateRange;
  mapboxMap: MapGetter;
  loading: number;
  errors: string[];
  // TODO this shouldn't be any
  layersData: LayerData<any>[];
  layerForms: LayerForm[];
};

// MapboxGL's map type contains some kind of cyclic dependency that causes an infinite loop in immers's change
// tracking. To save it off, we wrap it in a JS closure so that Redux just checks the function for changes, rather
// than recursively walking the whole object.
type MapGetter = () => MapBoxMap | undefined;

const initialState: MapState = {
  layers: [],
  dateRange: {} as DateRange,
  mapboxMap: (() => {}) as MapGetter,
  // Keep track of loading state with reference counting
  loading: 0,
  errors: [],
  layersData: [],
  layerForms: [],
};

export type FormInputChange = {
  layerId: string;
  inputId: string;
  value: string;
};

function keepLayer(layer: LayerType, payload: LayerType) {
  // Simple function to control which layers can overlap.
  return (
    payload.id !== layer.id &&
    (payload.type !== layer.type || payload.type === 'boundary')
  );
}

export const mapStateSlice = createSlice({
  name: 'mapState',
  initialState,
  reducers: {
    addLayer: (
      { layers, layerForms, ...rest },
      { payload }: PayloadAction<LayerType>,
    ) => {
      const layersToAdd = payload?.group?.activateAll
        ? Object.values(LayerDefinitions).filter(l =>
            payload?.group?.layers?.map(layer => layer.id).includes(l.id),
          )
        : [payload];

      const filteredLayers = layers.filter(layer => keepLayer(layer, payload));

      // Keep boundary layers at the top of our stack
      const newLayers =
        payload.type === 'boundary'
          ? [...layersToAdd, ...filteredLayers]
          : [...filteredLayers, ...layersToAdd];

      const layerForm =
        'formInputs' in payload
          ? { id: payload.id, inputs: payload.formInputs! }
          : null;

      return {
        ...rest,
        layers: newLayers,
        layerForms: layerForms.concat(layerForm ? [layerForm] : []),
      };
    },

    removeLayer: (
      { layers, ...rest },
      { payload }: PayloadAction<LayerType>,
    ) => ({
      ...rest,
      layers: layers.filter(({ id }) =>
        // Keep layers without group and layers with group from different group.
        payload.group
          ? !payload.group.layers.map(l => l.id).includes(id)
          : id !== payload.id,
      ),
    }),

    setFormInputValue: (
      { layerForms, ...rest },
      { payload }: PayloadAction<FormInputChange>,
    ) => ({
      ...rest,
      layerForms: layerForms.map(layerForm =>
        layerForm.id === payload.layerId
          ? {
              id: layerForm.id,
              inputs: layerForm.inputs.map(({ id, value, ...restInput }) => ({
                id,
                value: id === payload.inputId ? payload.value : value,
                ...restInput,
              })),
            }
          : layerForm,
      ),
    }),

    updateDateRange: (state, { payload }: PayloadAction<DateRange>) => ({
      ...state,
      dateRange: payload,
    }),

    setMap: (state, { payload }: PayloadAction<MapGetter>) => ({
      ...state,
      mapboxMap: payload,
    }),

    dismissError: (
      { errors, ...rest },
      { payload }: PayloadAction<string>,
    ) => ({
      ...rest,
      errors: errors.filter(msg => msg !== payload),
    }),

    updateLayerOpacity: (
      { layers, ...rest },
      { payload }: PayloadAction<Pick<LayerType, 'id' | 'opacity'>>,
    ) => ({
      ...rest,
      layers: layers.map(layer => ({
        ...layer,
        opacity: layer.id === payload.id ? payload.opacity : layer.opacity,
      })),
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      loadLayerData.fulfilled,
      (
        { layersData, loading, ...rest },
        { payload }: PayloadAction<LayerDataTypes>,
      ) => ({
        ...rest,
        loading: loading - 1,
        layersData: layersData.concat(payload),
      }),
    );

    builder.addCase(
      loadLayerData.rejected,
      ({ loading, errors, ...rest }, action) => ({
        ...rest,
        loading: loading - 1,
        errors: errors.concat(
          action.error.message ? action.error.message : action.error.toString(),
        ),
      }),
    );

    builder.addCase(loadLayerData.pending, ({ loading, ...rest }) => ({
      ...rest,
      loading: loading + 1,
    }));
  },
});

// Setters
export const {
  addLayer,
  removeLayer,
  setFormInputValue,
  updateDateRange,
  setMap,
  // TODO unused
  updateLayerOpacity,
} = mapStateSlice.actions;

export default mapStateSlice.reducer;
