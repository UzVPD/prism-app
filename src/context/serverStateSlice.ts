import { createSlice, PayloadAction, createAsyncThunk } from '@reduxjs/toolkit';
import { AvailableDates, KoboAuthParams } from '../config/types';
import { getLayersAvailableDates } from '../utils/server-utils';
import type { CreateAsyncThunkTypes, RootState } from './store';

type ServerState = {
  availableDates: AvailableDates;
  loading: boolean;
  error?: string;
  koboAuthParams?: KoboAuthParams;
};

const initialState: ServerState = {
  availableDates: {},
  loading: false,
};

export const loadAvailableDates = createAsyncThunk<
  AvailableDates,
  void,
  CreateAsyncThunkTypes
>('serverState/loadAvailableDates', () => getLayersAvailableDates());

export const serverStateSlice = createSlice({
  name: 'serverState',
  initialState,
  reducers: {
    updateLayersCapabilities: (
      state,
      { payload }: PayloadAction<AvailableDates>,
    ) => ({
      ...state,
      availableDates: payload,
    }),
    setLayerAccessToken: (
      state,
      { payload }: PayloadAction<KoboAuthParams>,
    ) => ({
      ...state,
      koboAuthParams: payload,
    }),
    clearKoboParams: state => ({
      ...state,
      koboAuthParams: undefined,
    }),
  },
  extraReducers: builder => {
    builder.addCase(
      loadAvailableDates.fulfilled,
      (state, { payload }: PayloadAction<AvailableDates>) => ({
        ...state,
        loading: false,
        availableDates: payload,
      }),
    );

    builder.addCase(loadAvailableDates.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAvailableDates.pending, state => ({
      ...state,
      loading: true,
    }));
  },
});

// Getters
export const availableDatesSelector = (
  state: RootState,
): ServerState['availableDates'] => state.serverState.availableDates;

export const isLoading = (state: RootState): ServerState['loading'] =>
  state.serverState.loading;

export const datesErrorSelector = (state: RootState): string | undefined =>
  state.serverState.error;

export const koboAuthParamsSelector = (
  state: RootState,
): KoboAuthParams | undefined => state.serverState.koboAuthParams;

// Setters
export const {
  updateLayersCapabilities,
  setLayerAccessToken,
  clearKoboParams,
} = serverStateSlice.actions;

export default serverStateSlice.reducer;
