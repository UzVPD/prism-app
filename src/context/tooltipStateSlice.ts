import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { merge } from 'lodash';
import type { RootState } from './store';

export interface PopupData {
  [key: string]: { data: number | string; coordinates: GeoJSON.Position };
}

export interface PopupComponentSpec {
  type: string;
  key: number;
  params: { [key: string]: any };
}

export interface PopupRemoteData {
  components: Array<PopupComponentSpec>;
}

export interface MapTooltipState {
  coordinates?: GeoJSON.Position;
  locationName: string;
  locationLocalName: string;
  data: PopupData;
  remoteData: PopupRemoteData | null;
  showing: boolean;
  wmsGetFeatureInfoLoading: boolean;
  remoteDataLoading: boolean;
}

type ShowPopupType = {
  coordinates: GeoJSON.Position;
  locationName: string;
  locationLocalName: string;
};

const initialState: MapTooltipState = {
  locationName: '',
  locationLocalName: '',
  data: {},
  remoteData: null,
  showing: false,
  wmsGetFeatureInfoLoading: false,
  remoteDataLoading: false,
};

export const fetchPopupData = createAsyncThunk(
  'tooltipState/fetchPopupData',
  async (url: string) => {
    return (
      await fetch(url, {
        mode: url.includes('http') ? 'cors' : 'same-origin',
      })
    ).json();
  },
);

export const tooltipStateSlice = createSlice({
  name: 'tooltipState',
  initialState,
  reducers: {
    addPopupData: (
      { data, ...rest },
      { payload }: PayloadAction<PopupData>,
    ) => ({
      ...rest,
      data: merge({}, data, payload),
    }),

    setPopupData: (state, { payload }: PayloadAction<PopupData>) => ({
      ...state,
      data: payload,
    }),

    clearRemotePopupData: state => ({
      ...state,
      remoteData: null,
    }),

    setPopupShowing: (state, { payload }: PayloadAction<boolean>) => ({
      ...state,
      showing: payload,
    }),

    hidePopup: state => ({
      ...state,
      showing: false,
      data: {},
    }),

    showPopup: (state, { payload }: PayloadAction<ShowPopupType>) => ({
      ...state,
      showing: true,
      locationName: payload.locationName,
      locationLocalName: payload.locationLocalName,
      coordinates: payload.coordinates,
    }),

    setPopupCoordinates: (
      state,
      { payload }: PayloadAction<GeoJSON.Position>,
    ) => ({
      ...state,
      coordinates: payload,
    }),

    setWMSGetFeatureInfoLoading: (
      state,
      { payload }: PayloadAction<boolean>,
    ) => ({
      ...state,
      wmsGetFeatureInfoLoading: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(fetchPopupData.fulfilled, (state, { payload }) => ({
      ...state,
      remoteData: payload,
      remoteDataLoading: false,
    }));

    builder.addCase(fetchPopupData.pending, state => ({
      ...state,
      remoteData: null,
      remoteDataLoading: true,
    }));

    builder.addCase(fetchPopupData.rejected, state => ({
      ...state,
      remoteData: null,
      remoteDataLoading: false,
    }));
  },
});

// Getters
export const tooltipSelector = (state: RootState): MapTooltipState =>
  state.tooltipState;
export const tooltipShowingSelector = (
  state: RootState,
): MapTooltipState['showing'] => state.tooltipState.showing;

// Setters
export const {
  addPopupData,
  setPopupData,
  clearRemotePopupData,
  setPopupShowing,
  hidePopup,
  showPopup,
  setPopupCoordinates,
  setWMSGetFeatureInfoLoading,
} = tooltipStateSlice.actions;

export default tooltipStateSlice.reducer;
