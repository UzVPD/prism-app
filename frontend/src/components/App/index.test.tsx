import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import App from '.';
import { store } from '../../context/store';

jest.mock('../NavBar', () => 'mock-NavBar');
jest.mock('../DataDrawer', () => 'mock-DataDrawer');
jest.mock('../MapView', () => 'mock-MapView');
jest.mock('../404Page', () => 'mock-NotFound');
jest.mock('../Notifier', () => 'mock-Notifier');
jest.mock('../AuthModal', () => 'mock-AuthModal');

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <App />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
