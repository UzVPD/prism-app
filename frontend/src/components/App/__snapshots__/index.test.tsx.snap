// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-notifier />
  <div
    id="app"
  >
    <mock-navbar
      isalertformopen="false"
    />
    <mock-mapview
      isalertformopen="false"
    />
    <mock-authmodal />
  </div>
</div>
`;
